### ISSUE TITLE
Failing to infer correct return type from partially stringified type annotation

### ISSUE THREAD

**Issue by bkis** [astral-sh/ty#3244]
### Summary

I encountered a problem where ty fails to correctly infer the return type when calling the `get` class method of a [Beanie ODM](https://github.com/BeanieODM/beanie) (v2.1.0) model. The implementation of the called method can be found [here](https://github.com/BeanieODM/beanie/blob/main/beanie/odm/documents.py#L219).

```py
from beanie import Document, PydanticObjectId

class Foo(Document):
    bar: str = "baz"

async def qux() -> None:
    foo = await Foo.get(PydanticObjectId("69d761e5b107b9e5f753515e"))  # Unknown | None
    print(foo.bar)  # Attribute `bar` is not defined on `None` in union `Unknown | None`
```

It's correct to error out when accessing `bar`, but the inferred type is wrong. For comparison, mypy correctly infers `Foo | None`, here.

### Version

ty 0.0.29

**Comment by sharkdp** [comment]
Thank you for reporting this.

It looks like ty has a problem with [this partially stringified `type["DocType"]` annotation](https://github.com/BeanieODM/beanie/blob/00c0f745ef12c4be145209d2ef69c2181d4d3a17/beanie/odm/documents.py#L220). When I change this to `cls: type[DocType]` or `cls: "type[DocType]"`, it works as expected.

It's not completely clear to me if this is a [legal type annotation](https://typing.python.org/en/latest/spec/annotations.html#type-and-annotation-expressions)?

**Comment by bkis** [comment]
I see, good catch! I also don't understand why it's necessary to use `"DocType"` instead of `DocType` here, as `DocType` is already defined before it's used. But this is a Beanie problem.

Regarding ty, if someone rules `type["DocType"]` valid, it might be a good idea to teach ty to handle it correctly. As I said, mypy does it, so there _is_ some reason to expect ty to do it, too :)

EDIT: `DocType` is bound to `Document` (`DocType = TypeVar("DocType", bound="Document")`) and thus cannot be referenced during the definition of `Document`, yet. So I guess stunts like `type["DocType"]` are actually needed, here.

**Comment by AlexWaygood** [comment]
> It's not completely clear to me if this is a [legal type annotation](https://typing.python.org/en/latest/spec/annotations.html#type-and-annotation-expressions)?

Hmm, yeah, it does look like a strict reading of the grammar would prohibit `type["DocType"]`, and would imply that you should fully stringify the annotation (`"type[DocType]"`) instead. I do think it makes sense for us to support string annotations in this location, though -- it's what users will probably expect (as demonstrated here), given that `list["ForwardRef"]` is clearly allowed by the grammar, and commonly used.

**Comment by sharkdp** [comment]
> I do think it makes sense for us to support string annotations in this location, though -- it's what users will probably expect (as demonstrated here), given that `list["ForwardRef"]` is clearly allowed by the grammar, and commonly used.

I agree. Also, all other type checkers seem to support `type["ForwardRef"]` as well.

**Comment by bkis** [comment]
Thanks for looking at this and replying so promptly!  
I really enjoy using ty – and the other dev-tools by astral. Once ty is completely up on its legs it will really fill a huge gap, as both a fast static type checker as well as a snappy, stable language server are nowhere else to be found.
